core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -165,10 +167,13 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
       // included in the shuffle write time.
       writeMetrics.incWriteTime(System.nanoTime() - openStartTime);
 
+      Set<Integer> nonEmptyPartitionSet = new HashSet<>();
       while (records.hasNext()) {
         final Product2<K, V> record = records.next();
         final K key = record._1();
-        partitionWriters[partitioner.getPartition(key)].write(key, record._2());
+        int partitionId = partitioner.getPartition(key);
+        partitionWriters[partitionId].write(key, record._2());
+        nonEmptyPartitionSet.add(partitionId);
```

Comment:
This is updating the Set for every tuple processed.
Instead, why not use `partitionWriterSegments[i].length > 0` instead in `writePartitionedData` ?